Move creating model instance function out from CoreDataStack mocks

crazytonyli opened the pull request:
This PR is part of paaHJt-3ky-p2.

## Changes
Two methods are deleted from `TestContextManager` and `ContextManagerMock`:
* `objectWithContentOfFile:`, replaced with new function in [`JSONLoader`](https://github.com/wordpress-mobile/WordPress-iOS/blob/6544502af7e1151e6c28ec768375d52ad3d1e557/WordPress/WordPressTest/TestUtilities/JSONLoader.swift)
* `loadEntityNamed:withContentsOfFile:`, replaced with a new extension function to `NSManagedObject`.

## Test Instructions
Make sure all unit tests pass.

## Regression Notes
N/A. All changes in this PR are in unit test, they don't have any impact on the app.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

crazytonyli committed:
Move creating model instance function out from CoreDataStack mocks

Two functions are moved out: one for parsing JSON file function and the
other for loading JSON file content into a Core Data model. Both aren't
related to managing a Core Data stack.

mokagio commented on WordPress/WordPressTest/TestUtilities/JSONLoader.swift:
```diff
@@ -38,6 +34,17 @@ import Foundation
+    ///
+    /// - Parameter filename: The full name of the json file to load.
+    /// - Returns: A dictionary representing the contents of the json file.
+    open func loadFile(named filename: String) -> JSONDictionary? {
```
Have you considered making this a `static` method? 

The current definition is consistent with the rest of the file, which is good. 

At the same time, the resulting usage is IMHO awkward:

```swift
JSONLoader().loadFile(named: fileName)
```

Why do we need to create a `JSONLoader` instance when there is no state access in the method (or in the class as a whole for the matter)?

Do you think it would be best to make the change right now, or to log this observation and address it with a self-contained wholesale `JSONLoader` update? _For what is worth, neither option is best in my view, just curious to hear your take on this_ 😄

crazytonyli replied:
Agreed. I don't think the `JSONLoader` instance is useful either. We could also add the JSON parsing function as an extension method to `JSONDictionary`, to save us from creating another arbitrary class `JSONLoader`, which isn't used in many places.

jkmassel replied:
If we're introducing a new static method anyway, I wonder if we could adjust the signature while we're at it?
```suggestion
    open func loadFile(named filename: String) throws -> JSONDictionary {
```

`XCTest` knows what to do with `throw`-ing functions, and the various `try?` calls here are swallowing errors that could make debugging easier 😅

WDYT?

crazytonyli replied:
Yes definitely. As I mentioned in [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r862666547) below, making this function throw was my initial thought, but the changes involved in various call sites made me go with the change that's more "compatible" with existing code.

But I went [a step further](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r863237278) to introduce a new pattern to load fixtures from the unit test bundle, I would appreciate your feedback on the new approach.

mokagio replied:
Thank @jkmassel! I don't know why, but I never reach for `throw` when thinking about API design in Swift, yet it's such a recommended and well-supported language-wise approach 😳 

As you mention, XCTest the framework has built-in handling for tests that throw at some point in the chain, so that sounds like a great approach. It would also be inline with what we do [in `ScreenObject`](https://github.com/Automattic/ScreenObject/blob/2d65beae47cdcaaf21703ce1b321f382fe0d0730/Sources/ScreenObject/ScreenObject.swift#L71) for the UI tests.

@crazytonyli [mentioned](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r862666547) the extra work involved in updating all the call stack to handle the `throws` and I can see how that might be daunting and would also result in a noisy PR diff. A suggestion that came to mind is to wrap this PR with a `fatalError` then follow up with one that replaces it with the throwing behavior.

mokagio commented on WordPress/WordPressTest/ActivityContentFactoryTests.swift:
```diff
@@ -15,6 +15,6 @@ final class ActivityContentFactoryTests: XCTestCase {
 
     private func getDictionaryFromFile(named fileName: String) -> [String: AnyObject] {
-        return contextManager.object(withContentOfFile: fileName) as! [String: AnyObject]
+        return JSONLoader().loadFile(named: fileName) ?? [:]
```
This nil coalescing is a good way to keep the compiler happy after the return type change.

How do you feel about returning an empty dictionary if the file is not found? Personally, I worry it might lead to test failing in non-straightforward ways if something happens to the file name or content.

crazytonyli replied:
Initially I made the new `JSONLoader` function and the new `NSManagedObjectContext` fixture method to throw, this change of course required all the call sites to keep throwing the error all the way up to test methods, which would help us diagnosing test failures. But I got lazy 😅, 'cause there are so many places need to change.

I think I'm going to do this properly this time - propagating errors from JSON parsing function all the way up to test methods, but also updating bunch of loading test data functions along the way, like [this one here](https://github.com/wordpress-mobile/WordPress-iOS/blob/112b352938ecbea191addd94be7f61d8cd4d5bf4/WordPress/WordPressTest/ActivityLogTestData.swift).

crazytonyli replied:
To optimise the error handling here and also [the loading JSON file function](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r862585396), I introduced [a new way](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478/files#diff-f3425ef339c7340197f30a8859ef98c5abfd92d8d13911940a4d426378c908b0) to create test objects from JSON files: https://github.com/wordpress-mobile/WordPress-iOS/pull/18478/commits/e9c69065efa4763e7681c04704d61d5b6b2847c8. I've update [notifications](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478/files#diff-4f9f4eadc7c4b2cb2a8677b7e9f9459fc70111e6e39228adb0cf40d3c299120c) related unit test to use this new pattern, to showcase how it's used in practice. If you think this is a good direction to go, I can delete the JSONLoader, and apply this new pattern to all unit tests, maybe in a separate PR, since there are going to be lots of changes.

mokagio replied:
> I introduced [a new way](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478/files#diff-f3425ef339c7340197f30a8859ef98c5abfd92d8d13911940a4d426378c908b0) to create test objects from JSON files 

Thank you for taking the time. I believe there's merit in the approach you're proposing, I particularly like the idea of a hierarchical name-spaced fixtures grab bag to help devs get the test data they need thanks to Xcode's autocompletion. **But I'm afraid I'll have to push back on this**.

Changing how we do something as foundational and widespread as loading fixture data is something we should approach with a lot of thoughtfulness. I'd like to see an RFC post for it, a code spike in line with what you did in https://github.com/wordpress-mobile/WordPress-iOS/commit/e9c69065efa4763e7681c04704d61d5b6b2847c8, and then an announcement post to let the product devs know about it.

Going back on your current proposal leaves the question of how to move forward to avoid tests failing in non-straightforward way 🤔  I left a [suggestion](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r863479696) in the thread about returning an `Optional` vs `throwing`.

crazytonyli replied:
Good call, starting with a RFC makes sense to me 👍 . I'll remove the fixture change out of this PR.

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+    /// - Parameters:
+    ///   - filename: The name of the JSON file to be loaded
+    ///   - context: The managed object context to use
+    /// - Returns: A new instance with property values of the given JSON file.
```
+1 for documenting using the Swift documentation standards, which results, among other things, in a nicely rendered help tooltip in Xcode:

<img width="839" alt="image" src="https://user-images.githubusercontent.com/1218433/166185891-a42de269-4064-4577-84f1-311a6b6ca619.png">

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+    ///   - filename: The name of the JSON file to be loaded
+    ///   - context: The managed object context to use
+    /// - Returns: A new instance with property values of the given JSON file.
+    static func fixture(fromFile filename: String, context: NSManagedObjectContext) -> Self {
```
Nitpick: Since "file name" is made up of two words, how do you feel about using the following camel cased name?

```suggestion
    static func fixture(fromFile fileName: String, context: NSManagedObjectContext) -> Self {
```

For what is worth, notice `fileName` is the naming used in a [method this PR touches](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478/files#diff-a2ec0dda4a3592793251f4f57621e00879f8710610411932210e7d2d98194ddaR17).

crazytonyli replied:
Sounds good to me. I'm not good at naming things at all, happy to following existing pattern!

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+    /// - Returns: A new instance with property values of the given JSON file.
+    static func fixture(fromFile filename: String, context: NSManagedObjectContext) -> Self {
+        guard let jsonObject = JSONLoader().loadFile(named: filename) else {
+            fatalError("Mockup data could not be parsed, the filename is \(filename)")
```
Have you considered the alternative of returning `nil` from here?

crazytonyli replied:
I kept the old behaviour which was crashing via `NSAssert`, but as mentioned in [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/18478#discussion_r862666547), I'll do it properly this time to make this function throw.

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+
+extension NSManagedObject {
+
+    /// Loads the contents of any given JSON file into a new NSManagedObject instance.
```
Nitpick: Just so documentation formats it as code

```suggestion
    /// Loads the contents of any given JSON file into a new `NSManagedObject` instance.
```

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+extension NSManagedObject {
+
+    /// Loads the contents of any given JSON file into a new NSManagedObject instance.
+    ///
+    /// This helper method is useful for Unit Testing scenarios.
```
How would you feel about removing this note?

After all, we are already in the unit test target 😄

mokagio commented on WordPress/WordPressTest/NSManagedObject+Fixture.swift:
```diff
@@ -0,0 +1,24 @@
+    ///   - filename: The name of the JSON file to be loaded
+    ///   - context: The managed object context to use
+    /// - Returns: A new instance with property values of the given JSON file.
+    static func fixture(fromFile filename: String, context: NSManagedObjectContext) -> Self {
```
Other nitpick: What do you think of using `insertInto context:`, like in [`init(entity:, insertInto:)`](https://developer.apple.com/documentation/coredata/nsmanagedobject)? I feel it would make it clear to readers what's the purpose of that parameter.

![image](https://user-images.githubusercontent.com/1218433/166187180-d24378a3-6e2b-434b-92a1-60b85b7e01f0.png)

crazytonyli replied:
Oh definitely. That's much nicer. 👍

mokagio commented on WordPress/WordPressTest/TestUtilities/JSONLoader.swift:
```diff
@@ -4,13 +4,12 @@ import Foundation
-    *
-    *  @returns    A dictionary representing the contents of the json file.
-    */
+    /// Loads the specified json file and returns a dictionary representing it.
+    ///
+    /// - Parameters:
+    ///   - name: The name of the file
+    ///   - type: The extension of the file
+    /// - Returns: A dictionary representing the contents of the json file.
```
Appreciate how you updated the documentation comments for consistency 👍

mokagio replied:
One small nitpick: In other parts of the PR, "json" appear all uppercase. As far as I understand, JSON is the appropriate way to write the word because it's an acronym.

```suggestion
    /// Loads the specified JSON file and returns a dictionary representing it.
    ///
    /// - Parameters:
    ///   - name: The name of the file
    ///   - type: The extension of the file
    /// - Returns: A dictionary representing the contents of the JSON file.
```

mokagio reviewed (commented):
Looking good so far @crazytonyli. Thank you for working on this!

I left a few questions, but mostly nitpicks. Keen to hear what you think.

---

One comment, which is on the edge between best practice and personal preference of mine. This PR does at least three things:

1. Extracts `loadEntityNamed:withContentsOfFile:` from `ContextManagerMock` into `NSManagedObject+Fixture` `fixture(fromFile:, context:)`
2. Extracts `objectWithContentOfFile:` from `ContextManagerMock` into `JSONLoader` `loadFile(named:)`
3. Updates some documentation comments formatting in `JSONLoader` to the Swift standard

All these changes occur in the same commit. **This is not a problem**. An alternative, though, could have been to make each operation in its own commit.

The value of that approach is that small commits are easy to reason about, and also to revert. It's also easier to cherry-pick a small commit from one branch to another (say if that's the only part of a PR that's worth keeping).

Anyway, this is just a little, mostly personal note I thought I'd share 😄 

---

Speaking of commits, I appreciate how you added a detailed message to yours 👍 

<img width="598" alt="image" src="https://user-images.githubusercontent.com/1218433/166187974-8e2b14fd-67bb-4252-9d11-e73a4324a2ee.png">

crazytonyli committed:
Rename argument name "filename" to "fileName"

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

crazytonyli committed:
Nice inline code format in function documentation

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

crazytonyli committed:
Update documentation: "json" -> "JSON"

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

crazytonyli committed:
Fix a compiling issue

crazytonyli committed:
Change argument label from `context` to `insertInto` for better API name

crazytonyli committed:
Replace JSONLoader with static functions

Currently fatalError is used to "handle errors", but they'll be replaced
with throwing errors later.

mokagio commented on WordPress/WordPressTest/TestUtilities/JSONLoader.swift:
```diff
@@ -1,51 +1,46 @@
-        }
-    }
+/// Class for finding the test bundler
+private class BundlerFinder {
```
I forgot to mention in the first pass: Thank you for making this `private` 😄

mokagio commented on WordPress/WordPressTest/TestUtilities/JSONLoader.swift:
```diff
@@ -1,51 +1,46 @@
-        } else {
-            return nil
-        }
+typealias JSONObject = Dictionary<String, AnyObject>
+
+extension JSONObject {
```
One last nitpick: Since the file now holds only `JSONObject`, how would you feel about renaming to `JSONObject.swift` for consistency?

crazytonyli replied:
Good catch! I missed that. It's renamed in latest commit.

mokagio reviewed (commented):
I run the tests locally with an invalid fixture name and they crashed as expected 👍 

<img width="1904" alt="Screen Shot 2022-05-04 at 10 29 38 am" src="https://user-images.githubusercontent.com/1218433/166608235-283587e6-fce1-48c0-86c0-c7fa012e46ff.png">

Left only one file name suggestion, then I'd say we're good to merge 👏

crazytonyli committed:
Rename file JSONLoader to JSONObject

mokagio reviewed (approved):
Thank you for following up. 

Ready to go :shipit: I updated #18487 to run CI. I'll also enable auto-merge so if CI is green this will get in.

crazytonyli commented:
@mokagio Thanks. After this PR is merged, I'll open a PR to make the JSON parsing function throw as discussed in previous comments.
